### PR TITLE
46 v1.2b BuyerList

### LINKED ISSUE #46: Make `BuyerList`
Model a list of `Buyer`s.

### PR DESCRIPTION
**Pull request by trxe**
closes #46.

Summary of changes:
- Made `Buyer` and `Property` implement a new interface called `Listable`
- Abstracted `UniquePropertyList` to `UniqueList`, from which `UniquePropertyList` and `UniqueBuyerList` now extends
- Similar changes for the related exceptions
- Refactored `UniquePropertyListTest` and created `UniqueBuyerListTest`

Also, is there a cleaner way to implement the `UniquePropertyList` and `UniqueBuyerList`, possibly by creating a kind of wrapper that can catch the exceptions `ListableNotFoundException` and `DuplicateListableException`?

@AY2122S1-CS2103T-W11-4/developers thank you

### CODE REVIEW

File: src/main/java/seedu/address/model/property/UniqueList.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.property.exceptions.DuplicateListableException;
+import seedu.address.model.property.exceptions.ListableNotFoundException;
+
+/**
+ * A abstract list that enforces uniqueness between its elements and does not allow nulls.
+ * A {@code Listable} item is considered unique by comparing using {@code Listable#isSameListable(Listable)}.
+ * As such, adding and updating of listables uses Listable#isSameListable(Listable) for equality
+ * so as to ensure that the property being added or updated is unique in terms of identity in the
```
Comment (sunjc826) [inline src/main/java/seedu/address/model/property/UniqueList.java:18]: ~`property`~ `element being added or updated`

File: src/main/java/seedu/address/model/property/UniqueList.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+ * UniqueListableList. However, the removal of a property uses Listable#equals(Object) so
```
Comment (sunjc826) [inline src/main/java/seedu/address/model/property/UniqueList.java:19]: Similarly, replace property with element. Ditto for line 20 as well (and other cases).

File: src/main/java/seedu/address/model/property/exceptions/DuplicateBuyerException.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package seedu.address.model.property.exceptions;
+
+/**
+ * Signals that the operation will result in duplicate Properties (Properties are considered duplicates
```
Comment (sunjc826) [inline src/main/java/seedu/address/model/property/exceptions/DuplicateBuyerException.java:4]: ~`Properties`~
`Buyers`

File: src/main/java/seedu/address/model/property/exceptions/DuplicateListableException.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package seedu.address.model.property.exceptions;
+
+/**
+ * Signals that the operation will result in duplicate Properties (Properties are considered duplicates
```
Comment (sunjc826) [inline src/main/java/seedu/address/model/property/exceptions/DuplicateListableException.java:4]: Properties -> elements

File: src/test/java/seedu/address/model/property/UniqueBuyerListTest.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+    private final UniqueBuyerList uniqueBuyerList = new UniqueBuyerList();
+
+    @Test
+    public void contains_nullBuyer_throwsNullPointerException() {
+        assertThrows(NullPointerException.class, () -> uniqueBuyerList.contains(null));
+    }
+
+    @Test
+    public void contains_propertyNotInList_returnsFalse() {
```
Comment (sunjc826) [inline src/test/java/seedu/address/model/property/UniqueBuyerListTest.java:30]: We should rename the methods from property to buyer as well.

**Review by sunjc826 — CHANGES_REQUESTED** [review]
The new abstractions are good! In-code documentation needs some renaming, however.

As for the conversion of `DuplicateListableException` to `Duplicate(Buyer|Property)Exception` as well as `ListableNotFoundException` to the more specific exception, this could be done via some reflective code, though I am not the most familiar with reflection in Java.
An example could be that you store the exception class e.g. `Class duplicateListableExceptionClazz = DuplicateBuyerException.class` as an *instance* variable tied to each list object. I do not think placing this as a static variable would work.

**Comment by trxe** [comment]
> 
> 
> The new abstractions are good! In-code documentation needs some renaming, however.
> 
> As for the conversion of `DuplicateListableException` to `Duplicate(Buyer|Property)Exception` as well as `ListableNotFoundException` to the more specific exception, this could be done via some reflective code, though I am not the most familiar with reflection in Java. An example could be that you store the exception class e.g. `Class duplicateListableExceptionClazz = DuplicateBuyerException.class` as an _instance_ variable tied to each list object and to throw the exception, do something like `throw new this.duplicateListableExceptionClazz`. (I barely know how the Class object works, so this is most likely inaccurate, but this is the gist.) I do not think placing this as a static variable would work.

Do you mean what is suggested here: https://stackoverflow.com/questions/6818251/is-it-wise-to-use-generics-for-exceptions/6818494#6818494

I think for now I'll just have the `Buyer` and `Property` exceptions extend the corresponding `Listable` exceptions, since implementing the exceptions like this might be unusual in this project.

**Review by sunjc826 — APPROVED** [review]
lgtm!

File: src/main/java/seedu/address/model/property/UniqueList.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+import java.util.Iterator;
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.property.exceptions.DuplicateListableException;
+import seedu.address.model.property.exceptions.ListableNotFoundException;
+
+/**
+ * A abstract list that enforces uniqueness between its elements and does not allow nulls.
```
Comment (lizchow) [inline src/main/java/seedu/address/model/property/UniqueList.java:15]: `UniqueList` is not a abstract class.
```suggestion
 * A generic list that enforces uniqueness between its elements and does not allow nulls.
```

File: src/main/java/seedu/address/model/property/exceptions/ListableNotFoundException.java [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package seedu.address.model.property.exceptions;
+
+/**
+ * Signals that the operation is unable to find the specified property.
```
Comment (lizchow) [inline src/main/java/seedu/address/model/property/exceptions/ListableNotFoundException.java:4]:
```suggestion
 * Signals that the operation is unable to find the specified listable.
```

File: src/main/java/seedu/address/model/property/exceptions/BuyerNotFoundException.java [hunk 8]
```diff
@@ -0,0 +1,6 @@
+package seedu.address.model.property.exceptions;
+
+/**
+ * Signals that the operation is unable to find the specified property.
+ */
+public class BuyerNotFoundException extends RuntimeException {}
```
Comment (lizchow) [inline src/main/java/seedu/address/model/property/exceptions/BuyerNotFoundException.java:6]: Should this extend `ListableNotFoundException`? Same for the `PropertyNotFoundException`.

File: src/main/java/seedu/address/model/property/UniqueList.java (hunk 6 above)
Comment (trxe) [reply src/main/java/seedu/address/model/property/UniqueList.java:15]: I forgot to change it after I decided to use generics instead, thanks for highlighting

File: src/main/java/seedu/address/model/property/exceptions/BuyerNotFoundException.java [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package seedu.address.model.property.exceptions;
+
+/**
+ * Signals that the operation is unable to find the specified property.
```
Comment (lizchow) [inline src/main/java/seedu/address/model/property/exceptions/BuyerNotFoundException.java:4]:
```suggestion
 * Signals that the operation is unable to find the specified buyer.
```

**Pull request merged**